### PR TITLE
Improve table search speed through lookups

### PR DESCRIPTION
**Pull request by indutny**
Prior to this change table search would have to do a binary search over about 1000 entries which resulted in around 10 memory loads on average. In this commit we reduce the search space by doing a pre-lookup in a generated table to get a smaller (often zero-length) slice of the full sorted range list. On average this gives us just one entry of the range list to perform binary search on, which reduces the average number of memory loads to 2.

### CODE REVIEW

**Comment by indutny** [comment]
With help of https://github.com/unicode-rs/unicode-segmentation/pull/113 this gives:

- Between 0% and 15% improvement on graphemes benchmark (depending on language)
- Between 22% and 40% improvement on word benchmarks (depending on language)

Full Report: https://gist.github.com/indutny/9f5f623c52d6e935225edfe09857ef99

**Comment by indutny** [comment]
Rebased over the latest master, but it looks like nothing changed in this PR.

File: src/tables.rs [hunk 1]
```diff
@@ -387,9 +387,81 @@ pub mod grapheme {
     }
 
     pub fn grapheme_category(c: char) -> (u32, u32, GraphemeCat) {
-        bsearch_range_value_table(c, grapheme_cat_table)
+        let idx = c as usize / 0x80;
```
Comment (Manishearth) [inline src/tables.rs:390]: please add/generate comments into the code that explains what's going on here

also on the const

Comment (indutny) [reply src/tables.rs]: Sounds good! Force pushed with comments!

**Comment by indutny-signal** [comment]
Let me run benchmarks on this first, it looks like I might have regressed with my latest refactor.

**Comment by Manishearth** [comment]
Alright, let me know when I should look at this!

**Comment by indutny** [comment]
@Manishearth  Should be good now, sorry for the back and forth! I made some simplifications and benchmarks looks even better than before: https://gist.github.com/indutny/c8d29a00680bfbaf19d22d02a7175c0d ( with up to 51% improvement on some word bounds benches).

**Review by Manishearth — APPROVED** [review]

**Comment by Manishearth** [comment]
Thanks!

**Comment by indutny** [comment]
Wow! Thank *you* for merging it!

Not to impose, but I was wondering when you planned to release a new version of the library?

**Comment by Manishearth** [comment]
I'd happily merge a PR bumping the library to 1.10.1.

**Comment by indutny** [comment]
Should it be 1.11.0 given the Unicode table update?

**Comment by jrose-signal** [comment]
Wait, I have one more optimization for you. :-)

**Comment by jrose-signal** [comment]
Ah, it looks like @indutny's optimization has made mine irrelevant—some texts get faster, others get slower. So it's probably not worth it.

**Pull request merged**